### PR TITLE
Support remaining functions in protobuf serialization, add `expr_fn` for `StructFunction`

### LINKED ISSUE #8098: Unsupported functions in protobuf serializatin
### Describe the bug

Inspired by https://github.com/apache/arrow-datafusion/pull/8089/files from @Syleechan  I checked if there were other functions not covered by our serialization logic. It turns out that yes there are:

### To Reproduce

Remove the catchall in parse_expr:
```diff
diff --git a/datafusion/proto/src/logical_plan/from_proto.rs b/datafusion/proto/src/logical_plan/from_proto.rs
index cdb0fe9bda..52a8480146 100644
--- a/datafusion/proto/src/logical_plan/from_proto.rs
+++ b/datafusion/proto/src/logical_plan/from_proto.rs
@@ -1645,9 +1645,6 @@ pub fn parse_expr(
                 )),
                 ScalarFunction::Isnan => Ok(isnan(parse_expr(&args[0], registry)?)),
                 ScalarFunction::Iszero => Ok(iszero(parse_expr(&args[0], registry)?)),
-                _ => Err(proto_error(
-                    "Protobuf deserialization error: Unsupported scalar function",
-                )),
             }
         }
         ExprType::ScalarUdfExpr(protobuf::ScalarUdfExprNode { fun_name, args }) => {
```

The compiler will tell you there are several more functions not covered *like `ToTimestamp` and `ArrowTypeOf`:

```
error[E0004]: non-exhaustive patterns: `datafusion::ScalarFunction::ToTimestamp`, `datafusion::ScalarFunction::StructFun`, `datafusion::ScalarFunction::ArrowTypeof` and 2 more not covered
    --> datafusion/proto/src/logical_plan/from_proto.rs:1303:19
     |
1303 |             match scalar_function {
     |                   ^^^^^^^^^^^^^^^ patterns `datafusion::ScalarFunction::ToTimestamp`, `datafusion::ScalarFunction::StructFun`, `datafusion::ScalarFunction::ArrowTypeof` and 2 more not covered
     |
note: `datafusion::ScalarFunction` defined here
```

What this means is that if a user tries to serialize a plan with `to_timestamp` or `arrow_typeof` it will error with "Protobuf deserialization error"

### Expected behavior

There should be no error

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JacobOgle**
## Which issue does this PR close?

Closes #8098.

## Rationale for this change

Suggested changes were made to catch most cases of datafusion::ScalarFunction

## What changes are included in this PR?

Added the additional cases for `ScalarFunction::ArrowTypeof`, `ScalarFunction::ToTimestamp`, 'ScalarFunction::Flatten'

## Are these changes tested?

Test suite seemed to pass.

## Are there any user-facing changes?

Not entirely sure of this. My first commit to this project so I am still getting up to speed.

None that I am aware of. Same reasoning as above answer.

### CODE REVIEW

**Review by Syleechan — APPROVED** [review]
good finding.

File: datafusion/proto/src/logical_plan/from_proto.rs [hunk 1]
```diff
@@ -1645,6 +1646,13 @@ pub fn parse_expr(
...
+                ScalarFunction::ArrowTypeof => {
+                    Ok(arrow_typeof(parse_expr(&args[0], registry)?))
+                }
+                ScalarFunction::ToTimestamp => {
+                    Ok(to_timestamp_seconds(parse_expr(&args[0], registry)?))
+                }
+                ScalarFunction::Flatten => Ok(flatten(parse_expr(&args[0], registry)?)),
                 _ => Err(proto_error(
                     "Protobuf deserialization error: Unsupported scalar function",
                 )),
```
Comment (alamb) [inline datafusion/proto/src/logical_plan/from_proto.rs:1658]: Can you please remove this catch all clause so the compiler will ensure all enum variants are covered?

```suggestion
```

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @JacobOgle  -- this is looking like a great start 🙏 

I started the CI to run the tests on this PR and I left a suggestion

File: datafusion/proto/src/logical_plan/from_proto.rs (hunk 1 above)
Comment (JacobOgle) [reply datafusion/proto/src/logical_plan/from_proto.rs:1658]: Not a problem! I will start chipping away at the last two enum variants here in a bit. Thanks for the feedback and openness allowing the contribution! 🙏

Comment (JacobOgle) [reply datafusion/proto/src/logical_plan/from_proto.rs:1662]: @alamb I was able to get the remaining cases covered with the exception to `datafusion::ScalarFunction::StructFun`. 

Browsing the code I am not seeing a supported function for this enum variant in `expr_fn.rs`. The scalar_expr macro maps enum variants in `built_in_functions.rs` but has no match for StructFun. Should this be mapped to the Struct variant in the macro?

Apologies for the confusion, but appreciate any guidance.

```
macro_rules! scalar_expr {
    ($ENUM:ident, $FUNC:ident, $($arg:ident)*, $DOC:expr) => {
        #[doc = $DOC ]
        pub fn $FUNC($($arg: Expr),*) -> Expr {
            Expr::ScalarFunction(ScalarFunction::new(
                built_in_function::BuiltinScalarFunction::$ENUM,
                vec![$($arg),*],
            ))
        }
    };
}
```

but only has
```
...
    // struct functions
    /// struct
    Struct,
...
```
in the `BuiltinScalarFunction::$ENUM`

**Review by alamb — APPROVED** [review]
Thank you @JacobOgle  and @Syleechan  -- this looks great -- a very nice improvement

**Pull request merged**